main.nf

```diff
@@ -2,83 +2,74 @@
-FOLD_INDEXES = K_FOLD_CELL_INDEXES.flatten().map{f -> tuple("${f}".split("\\.")[1], f) }
-
-// Data value channels 
-MATRIX_CH = Channel.fromPath(params.matrix)
-BARCODES_CH = Channel.fromPath(params.barcodes)
-FEATURES_CH = Channel.fromPath(params.features)
-SDRF_CH = Channel.fromPath(params.sdrf_processed)
+process generate_folds{
+    publishDir "${params.output_dir}", mode: 'copy'
+        conda "${baseDir}/envs/r-caret.yaml"
 
-//PROCESS NOT ITERATING!! WHY?
-// split data based on cell index folds
-	process split_train_test{
-	// We're hard coding the output dir in the config file	
-		publishDir "${params.output_dir}", mode: 'copy'
-		conda "${baseDir}/envs/r-caret.yaml"
-		
-		errorStrategy { task.exitStatus == 130 || task.exitStatus == 137  ? 'retry' : 'finish' }
-                maxRetries 5
-                memory { 16.GB * task.attempt }
+    errorStrategy { task.exitStatus == 130 || task.exitStatus == 137  ? 'retry' : 'finish' }
+        maxRetries 5
+        memory { 16.GB * task.attempt }
+    
+    input:
+        file(matrix) from MATRIX_CH        
+    
+    output:
+        file("folds/*") into K_FOLD_CELL_INDEXES 
 
-                input:
-		tuple val(fold), file(k_fold_cell_indexes) from FOLD_INDEXES 
-		file(matrix) from MATRIX_CH 
-		file(barcodes) from BARCODES_CH 
-		file(features) from FEATURES_CH 
-		file(sdrf) from SDRF_CH
-		
-		output:
-		tuple val(fold), file("test.zip"), file("train.zip") into SPLIT_DATA 
-		
-		"""
-                rm -rf ${params.split_train_test.output_test_dir} && mkdir ${params.split_train_test.output_test_dir}
-                rm -rf ${params.split_train_test.output_train_dir} && mkdir ${params.split_train_test.output_train_dir}
-                #mkdir ${params.split_train_test.output_train_dir}
-
-		split-train-test-data.R\
-                        --input-matrix ${matrix}\
-                        --input-cell-indexes ${k_fold_cell_indexes}\
-			--input-barcodes-tsv ${barcodes}\
-			--input-features-tsv ${features}\
-			--input-processed-sdrf ${sdrf}\
-                        --output-test-dir ${params.split_train_test.output_test_dir}\
-                        --output-train-dir ${params.split_train_test.output_train_dir}
-		#zip files by test or train
-		zip -r test.zip ${params.split_train_test.output_test_dir}
-		zip -r train.zip ${params.split_train_test.output_train_dir}
-                """
-        }
+    script:
+    """
+        zipdir=\$(unzip -qql $matrix | head -n1 | tr -s ' ' | cut -d' ' -f5- | sed 's|/||')
+        unzip -p $matrix \${zipdir}/barcodes.tsv > barcodes.tsv
+        
+        caret-create-folds.R\
+                --input-barcodes-tsv barcodes.tsv\
+                --k-folds-number ${params.generate_folds.k_folds_num}\
+                --output-dir folds
+        """
+}
 
+K_FOLD_CELL_INDEXES
```

Comment:
So, flatten here takes things from:

```
[ 'foo/1', 'foo/2', 'foo/3']
```
... to:

```
[ 'foo/1' ]
[ 'foo/2' ]
[ 'foo/3' ]
```
Then the map function extracts the values from the file names to create tuples:

```
[ '1', 'foo/1' ]
..etc
```